src/TokenHandler/AccountManager.mo

```diff
@@ -0,0 +1,408 @@
+
+      let latestDeposit = try {
+        await* loadDeposit(p);
+      } catch (err) {
+        depositRegistry.unlock(p);
+        throw err;
+      };
+
+      depositRegistry.unlock(p);
+
+      let prevDeposit = updateDeposit(p, latestDeposit);
+
+      if (latestDeposit < prevDeposit) freezeCallback("latestDeposit < prevDeposit on notify");
+
+      if (latestDeposit > fee_) {
+        // precredit deposit funds
+        ignore creditRegistry.credit(p, latestDeposit - fee_);
+        // schedule consolidation for this p
+        pushToBacklog(p, latestDeposit);
+        // schedule a canister self-call to process the backlog
+        // we need try-catch so that we don't trap if scheduling fails synchronously
+        try ignore processBacklog() catch (_) {};
+      };
+
+      let depositDelta = latestDeposit - prevDeposit : Nat;
+
+      if (depositDelta > 0) journal.push((Time.now(), p, #newDeposit(depositDelta)));
+
+      return ?depositDelta;
+    };
+
+    /// Processes the consolidation transfer for a principal.
+    func processConsolidationTransfer(p : Principal, deposit : Nat) : async* {
+      #Ok : Nat;
+      #Err : ICRC1.TransferError or {
+        #CallIcrc1LedgerError;
+        #InsufficientDeposit;
+      };
+    } {
+      if (deposit <= fee_) return #Err(#InsufficientDeposit);
+
+      let transferAmount : Nat = Int.abs(deposit - fee_);
+
+      let transferResult = try {
+        await icrc1Ledger.icrc1_transfer({
+          from_subaccount = ?Mapping.toSubaccount(p);
+          to = { owner = ownPrincipal; subaccount = null };
+          amount = transferAmount;
+          fee = ?fee_;
+          memo = null;
+          created_at_time = null;
+        });
+      } catch (_) {
+        #Err(#CallIcrc1LedgerError);
+      };
+
+      switch (transferResult) {
+        case (#Ok _) {
+          ignore updateDeposit(p, 0);
+          totalConsolidated_ += transferAmount;
+          journal.push((Time.now(), p, #consolidated({ deducted = deposit; credited = transferAmount })));
+        };
+        case (#Err err) {
+          journal.push((Time.now(), p, #consolidationError(err)));
+        };
+      };
+
+      transferResult;
+    };
+
+    /// Attempts to consolidate the funds for a particular principal.
+    func consolidate(p : Principal) : async* () {
+      let latestDeposit = try { await* loadDeposit(p) } catch (err) {
+        pushToBacklog(p, 0);
+        return;
+      };
+
+      ignore updateDeposit(p, latestDeposit);
+
+      ignore depositRegistry.set(
+        p,
+        func(info) {
+          info.underway := latestDeposit;
+          underwayFunds += latestDeposit - info.underway;
+          true;
+        },
+      );
+
+      let transferResult = await* processConsolidationTransfer(p, latestDeposit);
+
+      switch (transferResult) {
+        case (#Err(#InsufficientDeposit)) {
+          ignore creditRegistry.debit(p, latestDeposit - fee_);
```

Comment:
This line should not be here, or? 

#InsufficientDeposit means latestDeposit <= fee_. We should have taken care of the correct credit amount in notify().

If this happened due to a fee change then we would end up in in #BadFee and should leave debit/credit to reprocessing the backlog.

Reply:
I agree that we should not correct credit funds when consolidation failed and in this case we just repush the deposit to the backlog. This is a mechanical error, I will fix it

But in case of insufficient deposit we should correct credited deposit funds, because they are precredited in notify(). Same when bad fee, we just correct credited funds based on actual fee.

Reply:
So in the case #InsufficientDeposit we detect dust deposit (that was not dust when pushed in the backlog) and debit it. Because user cannot spend this actually

Reply:
But #InsufficientDeposit cannot really occur, can it? We can only get to this line if we have a bug, or?